jax/_src/lax/lax.py

```diff
@@ -1831,13 +1831,53 @@ def logistic_impl(x):
 mlir.register_lowering(logistic_p,
                        mlir.lower_fun(logistic_impl, multiple_results=False))
 
+sin_f_p = standard_unop(_float, 'sin_f')
+ad.defjvp(sin_f_p, lambda g, x: mul(g, cos(x)))
+mlir.register_lowering(sin_f_p, partial(_nary_lower_hlo, hlo.sine))
+
+cos_f_p = standard_unop(_float, 'cos_f')
+ad.defjvp(cos_f_p, lambda g, x: neg(mul(g, sin(x))))
+mlir.register_lowering(cos_f_p, partial(_nary_lower_hlo, hlo.cosine))
```

Comment:
I think we should try to do this without introducing a new primitive – there are various custom transformations that people use that may be tripped up by the introduction of this.

A pattern I've used for this in the past might look something like this: define `sin_complex` in Python as below, and wrap it in `mlir.lower_fun` then write a lowering rule that checks the dtype and either calls the lowered `sin_impl`, or calls `hlo.sine` directly. Then define the impl rule via `dispatch.simple_impl`.

What do you think? Feel free to let me know if you have questions – we could also schedule a half hour to chat and/or paircode if you think it would be helpful. Thanks!

Reply:
@jakevdp , thanks for the hints! I have applied the suggested pattern except I didn't use `dispatch.simple_impl` because I am not sure about its relevance, so suggestions are welcome.

Reply:
Oh, I guess `standard_unop` takes care of the impl for you.